### ISSUE TITLE
String datatype in HDF5

### ISSUE THREAD

**Issue by ajtritt** [NeurodataWithoutBorders/pynwb#13]
Originally reported by: **Oliver Ruebel (Bitbucket: [oruebel](https://bitbucket.org/oruebel), GitHub: [oruebel](https://github.com/oruebel))**

----------------------------------------

pynwb.io.h5tools .__get_type specifies  _h5py.special_dtype(vlen=bytes) as datatype for string objects. It looks like this implies: 1) one cannot store non-ASCII characters, and 2) when loading the string dataset we get back a bytestring so that one has to call decode('utf-8') first to get a unicode string and a user needs to know the encoding. In Python2 h5py used different datatype for variable length ASCII and unicode strings. Is this no longer supported?

----------------------------------------
- Bitbucket: https://bitbucket.org/lblneuro/pynwb/issue/13

**Comment by ajtritt** [comment]
*Original comment by* **Andrew Tritt (Bitbucket: [ajtritt](https://bitbucket.org/ajtritt), GitHub: [ajtritt](https://github.com/ajtritt))**:

----------------------------------------

I think that was a hack to get strings as datasets/attributes to validate with the original NWB API. The validator is in Python2, where reading in Unicode is not default. 

You are correct on the implications. Yes, Python3 still supports the ability to store Unicode vs ASCII differently

I am unclear about what the bug is, since no behavioral need was ever defined. We should decide on a behavior and implement the solution needed for that. My goal is to maintain simplicity. I do not want users to have to think about storing things as ASCII or Unicode. Since we are using Python 3, and strings are unicode by default, we can force storage of unicode in HDF5. This, however, has implications for future APIs. Things are fine if one stays in Python3, but if they leave Python3 where unicode is not default, it could complicate other developer's efforts (I am not sure how much we should focus on this). 

In Python, it is trivial to read ASCII in as Unicode if we don't allow string datasets (which I think we should do anyway), but I am perfectly fine forcing the world of NWB developers to support Unicode if we think users will want to store unicode strings.

**Comment by ajtritt** [comment]
*Original comment by* **Oliver Ruebel (Bitbucket: [oruebel](https://bitbucket.org/oruebel), GitHub: [oruebel](https://github.com/oruebel))**:

----------------------------------------

I guess bug may be the wrong classification for this item. I ran into this issue while writing a test case for the iterative data write where I stored a unicode string and then the assert that compared the original string with the string that was loaded from file failed because the input string was unicode and the other was a byte string. 

In general, I think it will be a good idea to have unicode as a default, otherwise any user outside of the US will be in trouble if they want to store strings in their language. From a write perspective it should not matter for users. It's really only an issue on read if someone else's code does not support Unicode on read. 

If we want to support both ASCII and Unicode options, then maybe that should have an option on the HDF5 write part, where by default we would use unicode and only switch to ASCII if the user specifically requests it. On the user-facing Python API side I think it's fine to use unicode.
